File: leetcode/solutions/0202.HappyNumber/isHappy.go
package solution

func isHappy(n int) bool {
	numMap := map[int]bool{}

	for !numMap[n] {
		// Just to break infinite loop. Never be read by below
		numMap[n] = true

		sum := 0

		for ; n > 0; n /= 10 {
			num := n % 10
			sum += num * num
		}

		n = sum
	}

	return n == 1
}

File: leetcode/solutions/0202.HappyNumber/isHappy.ts
export default function isHappy(n: number): boolean {
  const set = new Set<number>();

  while (!set.has(n)) {
    // To break inifinite loop.
    set.add(n);

    let sum = 0;

    while (n > 0) {
      sum += (n % 10) ** 2;
      n = Math.floor(n / 10);
    }

    n = sum;
  }

  return n === 1;
}
